### PR TITLE
allow package.json to be located in a parent directory of the pulumi …

### PR DESCRIPTION
**Pull request by RobbieMcKinstry**
Continuing to have issues with CI, I've once again pushed this PR to a branch to attempt to pass status checks.

Original PR: https://github.com/pulumi/pulumi/pull/11045

### CODE REVIEW

**Review by dixler — APPROVED** [review]

**Comment by dixler** [comment]
Few errors that look like
```
Diagnostics:
  pulumi:pulumi:Stack (esm-js-p-it-fv-az163-4-esm-js-c6787da6):
    error: Running program '/tmp/p-it-fv-az163-4-esm-js-c6787da6-3886097518' failed with an unhandled exception:
    Error [ERR_REQUIRE_ESM]: require() of ES Module /home/runner/work/pulumi/pulumi/sdk/nodejs/bin/node_modules/pkg-dir/index.js from /home/runner/work/pulumi/pulumi/sdk/nodejs/bin/cmd/run/run.js not supported.
    Instead change the require of index.js in /home/runner/work/pulumi/pulumi/sdk/nodejs/bin/cmd/run/run.js to a dynamic import() which is available in all CommonJS modules.
        at /home/runner/work/pulumi/pulumi/sdk/nodejs/bin/cmd/run/run.js:69:107
```

**Comment by RobbieMcKinstry** [comment]
Yeah, i thought this was fixed in the original PR, maybe I'm a commit behind somehow? I can cherry-pick the commit directly.

**Comment by RobbieMcKinstry** [comment]
WOW! After a serious battle with our test suite, I found the issue. The original PR should check if the path to the Pulumi program is a directory, and start there instead of the parent if it is.

File: sdk/nodejs/cmd/run/run.ts [hunk 1]
```diff
@@ -43,13 +54,35 @@ function reportModuleLoadFailure(program: string, error: Error): never {
...
+async function npmPackageRootFromProgramPath(programPath: string): Promise<string> {
+    // pkgDir is an ESM module which we use to find the location of package.json
+    // Because it's an ESM module, we cannot import it directly.
+    const pkgDir = await dynamicImport('pkg-dir');
+    // Check if programPath is a directory. If not, then we
+    // look at it's parent dir for the package root.
+    let isDirectory = false;
+    if(fs.existsSync(programPath)) {
+        const fileStat = await fspromises.lstat(programPath);
+        isDirectory = fileStat.isDirectory();
```
Comment (dixler) [inline sdk/nodejs/cmd/run/run.ts:75]: I feel like programPath not existing would merit a short-circuit no?

File: sdk/nodejs/cmd/run/run.ts [hunk 2]
```diff
@@ -43,13 +54,35 @@ function reportModuleLoadFailure(program: string, error: Error): never {
...
     if (stat.isDirectory()) {
-        return program;
+        return programPath;
     } else {
-        return path.dirname(program);
+        return path.dirname(programPath);
+    }
+}
+
+async function npmPackageRootFromProgramPath(programPath: string): Promise<string> {
```
Comment (dixler) [inline sdk/nodejs/cmd/run/run.ts:66]: :+1: for more precise naming

**Review by dixler — COMMENTED** [review]
Adding the dependency looks good since we had a discussion on the previous PR about adding it.

Sadly the linter is complaining so we'll need to fix that.

Had a question about a file exists check. Other than that things are looking good.

File: sdk/nodejs/cmd/run/run.ts (hunk 2 above)
Comment (RobbieMcKinstry) [reply sdk/nodejs/cmd/run/run.ts:66]: Credit goes to @geekflyer for that one!

**Comment by RobbieMcKinstry** [comment]
@dixler since you marked the PR as approved, I'm assuming your feedback about early exit is nonblocking. I can add a follow-on PR if you disagree with my reply 👍🏻  :D 

Thanks very much for your help with this one!!!

**Comment by RobbieMcKinstry** [comment]
This might get kicked out of queue by the failing integ test.

**Comment by RobbieMcKinstry** [comment]
ugh, i can't reproduce the failing test locally. Integ tests are broken on my machine.

**Comment by RobbieMcKinstry** [comment]
Looks like two path segments were reversed in the test case. Going to throw this at CI since I can't repro locally... :(

**Comment by RobbieMcKinstry** [comment]
Cool, the test runs but fails, so there's another bug in the program logic.

**Comment by RobbieMcKinstry** [comment]
I'm pretty sure [this test never worked](https://github.com/pulumi/pulumi/pull/11876/files#diff-e26b69e905f24f969a12360b000fa75a1881737f03b4cf99d44585992e5b40ccR1121-R1128) (correctly, at least!). The integration tests copy over the program code only, never the `package.json` file that lives in the directory above, and is required to validate the feature.

Without pretty heavily changing our integration test framework, I won't be able to tweak this code to get the test to execute correctly. Looking into alternative methods to test.

**Comment by RobbieMcKinstry** [comment]
Closed in favor of https://github.com/pulumi/pulumi/pull/12759

**Pull request closed without merging**